src/hotspot/cpu/aarch64/abstractInterpreter_aarch64.cpp

```diff
@@ -165,6 +165,19 @@ void AbstractInterpreter::layout_activation(Method* method,
     popframe_extra_args;
   interpreter_frame->interpreter_frame_set_last_sp(esp);
 
+  // We have to add extra reserved slots to max_stack. There are 3 users of the extra slots,
+  // none of which are at the same time, so we just need to make sure there is enough room
+  // for the biggest user:
+  //   -reserved slot for exception handler
+  //   -reserved slots for JSR292. Method::extra_stack_entries() is the size.
+  //   -reserved slots for TraceBytecodes
+  int max_stack = method->constMethod()->max_stack() + MAX2(3, Method::extra_stack_entries());
+  intptr_t* extended_sp = (intptr_t*) monbot  -
+    (max_stack * Interpreter::stackElementWords) -
+    popframe_extra_args;
+  extended_sp = align_down(extended_sp, StackAlignmentInBytes);
+  interpreter_frame->interpreter_frame_set_extended_sp(extended_sp);
+
   // All frames but the initial (oldest) interpreter frame we fill in have
   // a value for sender_sp that allows walking the stack but isn't
   // truly correct. Correct the value here.
```

Comment:
The condition this comment leads into tests for
```
interpreter_frame->sender_sp() == interpreter_frame->interpreter_frame_sender_sp())
```
I'm not really clear how/where that actually happens? Can you explain?

Reply:
Sure.
First thing to realize is that this method is used only by deopt. The logic here removes the caller's args that it's pushed onto the stack, iff the caller is an interpreted method.
I think this is a hangover from x86. It doesn't hurt, but neither does it do anything useful AFAICS because we're going to restore the caller's SP from `interpreter_frame_extended_sp` and the caller's ESP from `interpreter_frame_last_sp` as soon as we return.
I guess it's possible that something somewhere that walks the stack depends on this logic, but if it does I can't find it.

Reply:
Ah yes, I should have looked at the calling code. Ta!